lib/rucio/core/rule.py

```diff
@@ -143,10 +143,10 @@ def default(rule: models.ReplicationRule, did: models.DataIdentifier, session: '
 
 
 @transactional_session
-def add_rule(dids, account, copies, rse_expression, grouping, weight, lifetime, locked, subscription_id,
-             source_replica_expression=None, activity='User Subscriptions', notify=None, purge_replicas=False,
-             ignore_availability=False, comment=None, ask_approval=False, asynchronous=False, ignore_account_limit=False,
-             priority=3, delay_injection=None, split_container=False, meta=None, *, session: "Session", logger=logging.log):
+def add_rule(dids: Sequence[dict[str, Any]], account: InternalAccount, copies: int, rse_expression: str, grouping: Literal['ALL', 'DATASET', 'NONE'], weight: str, lifetime: int, locked: bool, subscription_id: str,
+             source_replica_expression: Optional[str] = None, activity: str = 'User Subscriptions', notify: Optional[Literal['Y', 'N', 'C']] = None, purge_replicas: bool = False,
+             ignore_availability: bool = False, comment: Optional[str] = None, ask_approval: bool = False, asynchronous: bool = False, ignore_account_limit: bool = False,
+             priority: int = 3, delay_injection: Optional[int] = None, split_container: bool = False, meta: Optional[dict[str, Any]] = None, *, session: "Session", logger: LoggerFunction = logging.log) -> list[str]:
```

Comment:
In addition, `weight` can be `None`. Also for `RuleDict`.

Reply:
1. I agree with you and I've now made this change for the function signatures that were too long, but I would argue that until we have explicit style guidelines (https://github.com/rucio/documentation/issues/287) and checks in the CI that enforce these guidelines, the only style that can be inferred is the `pre-existing` one in the file; in this case it wasn't one parameter per line, but I think it might have been based on line length.